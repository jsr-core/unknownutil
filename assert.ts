import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
