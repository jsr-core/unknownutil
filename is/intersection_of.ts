import { rewriteName } from "../_funcutil.ts";
import { hasAnnotation, type WithPredObj } from "../_annotation.ts";
import type { Predicate } from "../type.ts";
import { isObjectOf } from "./object_of.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` is `IntersectionOf<T>`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const isMyType = is.IntersectionOf([
 *   is.ObjectOf({ a: is.Number }),
 *   is.ObjectOf({ b: is.String }),
 * ]);
 * const a: unknown = { a: 0, b: "a" };
 * if (isMyType(a)) {
 *   const _: { a: number } & { b: string } = a;
 * }
 * ```
 *
 * Depending on the version of TypeScript and how values are provided, it may be necessary to add `as const` to the array
 * used as `preds`. If a type error occurs, try adding `as const` as follows:
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const preds = [
 *   is.ObjectOf({ a: is.Number }),
 *   is.ObjectOf({ b: is.String }),
 * ] as const
 * const isMyType = is.IntersectionOf(preds);
 * const a: unknown = { a: 0, b: "a" };
 * if (isMyType(a)) {
 *   const _: { a: number } & { b: string } = a;
 * }
 * ```
 */
export function isIntersectionOf<
  T extends readonly [
    Predicate<unknown> & WithPredObj<Record<PropertyKey, Predicate<unknown>>>,
    ...(
      & Predicate<unknown>
      & WithPredObj<Record<PropertyKey, Predicate<unknown>>>
    )[],
  ],
>(
  preds: T,
):
  & Predicate<IntersectionOf<T>>
  & WithPredObj<Record<PropertyKey, Predicate<unknown>>>;
export function isIntersectionOf<
  T extends readonly [Predicate<unknown>],
>(
  preds: T,
): T[0];
export function isIntersectionOf<
  T extends readonly [Predicate<unknown>, ...Predicate<unknown>[]],
>(
  preds: T,
):
  & Predicate<IntersectionOf<T>>
  & WithPredObj<Record<PropertyKey, Predicate<unknown>>>;
export function isIntersectionOf<
  T extends readonly [Predicate<unknown>, ...Predicate<unknown>[]],
>(
  preds: T,
):
  | Predicate<unknown>
  | Predicate<IntersectionOf<T>>
    & WithPredObj<Record<PropertyKey, Predicate<unknown>>> {
  const predObj = {};
  const restPreds = preds.filter((pred) => {
    if (!hasAnnotation(pred, "predObj")) {
      return true;
    }
    Object.assign(predObj, pred.predObj);
  });
  if (restPreds.length < preds.length) {
    restPreds.push(isObjectOf(predObj));
  }
  if (restPreds.length === 1) {
    return restPreds[0];
  }
  return rewriteName(
    (x: unknown): x is IntersectionOf<T> => restPreds.every((pred) => pred(x)),
    "isIntersectionOf",
    preds,
  );
}

type TupleToIntersection<T> = T extends readonly [] ? never
  : T extends readonly [infer U] ? U
  : T extends readonly [infer U, ...infer R] ? U & TupleToIntersection<R>
  : never;

type IntersectionOf<T> = TupleToIntersection<
  {
    -readonly [P in keyof T]: T[P] extends Predicate<infer U> ? U : never;
  }
>;
