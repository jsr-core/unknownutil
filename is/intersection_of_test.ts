import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType, type IsExact } from "@std/testing/types";
import { is } from "./mod.ts";
import { isIntersectionOf } from "./intersection_of.ts";

Deno.test("isIntersectionOf<T>", async (t) => {
  const objPreds = [
    is.ObjectOf({ a: is.Number }),
    is.ObjectOf({ b: is.String }),
  ] as const;
  const mixPreds = [
    is.Function,
    is.ObjectOf({ b: is.String }),
  ] as const;

  await t.step("returns properly named predicate function", async (t) => {
    assertEquals(typeof isIntersectionOf([is.String]), "function");
    await assertSnapshot(t, isIntersectionOf([is.String]).name);
    await assertSnapshot(t, isIntersectionOf(objPreds).name);
    await assertSnapshot(t, isIntersectionOf(mixPreds).name);
  });

  await t.step("returns true on all of T", () => {
    const f = Object.assign(() => void 0, { b: "a" });
    assertEquals(isIntersectionOf([is.String])("a"), true);
    assertEquals(isIntersectionOf(objPreds)({ a: 0, b: "a" }), true);
    assertEquals(isIntersectionOf(mixPreds)(f), true);
  });

  await t.step("returns false on non of T", () => {
    const f = Object.assign(() => void 0, { b: "a" });
    assertEquals(isIntersectionOf(objPreds)("a"), false);
    assertEquals(isIntersectionOf(mixPreds)({ a: 0, b: "a" }), false);
    assertEquals(isIntersectionOf([is.String])(f), false);
  });

  await t.step("predicated type is correct", () => {
    const a: unknown = undefined;

    if (isIntersectionOf([is.String])(a)) {
      assertType<IsExact<typeof a, string>>(true);
    }

    if (isIntersectionOf(objPreds)(a)) {
      assertType<IsExact<typeof a, { a: number } & { b: string }>>(true);
    }

    if (isIntersectionOf(mixPreds)(a)) {
      assertType<
        IsExact<
          typeof a,
          & ((...args: unknown[]) => unknown)
          & { b: string }
        >
      >(true);
    }
  });

  await t.step("predicated type is correct (#68)", () => {
    const a: unknown = undefined;
    const pred = isIntersectionOf([
      is.ObjectOf({ id: is.String }),
      is.UnionOf([
        is.ObjectOf({ result: is.String }),
        is.ObjectOf({ error: is.String }),
      ]),
    ]);

    if (pred(a)) {
      assertType<
        IsExact<
          typeof a,
          { id: string } & ({ result: string } | { error: string })
        >
      >(true);
    }
  });

  await t.step("with symbol properties", async (t) => {
    const b = Symbol("b");
    const objPreds = [
      is.ObjectOf({ a: is.Number }),
      is.ObjectOf({ [b]: is.String }),
    ] as const;
    const mixPreds = [
      is.Function,
      is.ObjectOf({ [b]: is.String }),
    ] as const;

    await t.step("returns properly named predicate function", async (t) => {
      assertEquals(typeof isIntersectionOf([is.String]), "function");
      await assertSnapshot(t, isIntersectionOf([is.String]).name);
      await assertSnapshot(t, isIntersectionOf(objPreds).name);
      await assertSnapshot(t, isIntersectionOf(mixPreds).name);
    });

    await t.step("returns true on all of T", () => {
      const f = Object.assign(() => void 0, { [b]: "a" });
      assertEquals(isIntersectionOf([is.String])("a"), true);
      assertEquals(isIntersectionOf(objPreds)({ a: 0, [b]: "a" }), true);
      assertEquals(isIntersectionOf(mixPreds)(f), true);
    });

    await t.step("returns false on non of T", () => {
      const f = Object.assign(() => void 0, { [b]: "a" });
      assertEquals(isIntersectionOf(objPreds)("a"), false);
      assertEquals(isIntersectionOf(mixPreds)({ a: 0, [b]: "a" }), false);
      assertEquals(isIntersectionOf([is.String])(f), false);
    });

    await t.step("predicated type is correct", () => {
      const a: unknown = undefined;

      if (isIntersectionOf([is.String])(a)) {
        assertType<IsExact<typeof a, string>>(true);
      }

      if (isIntersectionOf(objPreds)(a)) {
        assertType<IsExact<typeof a, { a: number } & { [b]: string }>>(true);
      }

      if (isIntersectionOf(mixPreds)(a)) {
        assertType<
          IsExact<
            typeof a,
            & ((...args: unknown[]) => unknown)
            & { [b]: string }
          >
        >(true);
      }
    });

    await t.step("predicated type is correct (#68)", () => {
      const a: unknown = undefined;
      const pred = isIntersectionOf([
        is.ObjectOf({ id: is.String }),
        is.UnionOf([
          is.ObjectOf({ result: is.String }),
          is.ObjectOf({ error: is.String }),
        ]),
      ]);

      if (pred(a)) {
        assertType<
          IsExact<
            typeof a,
            { id: string } & ({ result: string } | { error: string })
          >
        >(true);
      }
    });
  });
});
