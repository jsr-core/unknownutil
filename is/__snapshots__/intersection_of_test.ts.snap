export const snapshot = {};

snapshot[`isIntersectionOf<T> > returns properly named predicate function 1`] = `"isString"`;

snapshot[`isIntersectionOf<T> > returns properly named predicate function 2`] = `
"isObjectOf({
  a: isNumber,
  b: isString
})"
`;

snapshot[`isIntersectionOf<T> > returns properly named predicate function 3`] = `
"isIntersectionOf([
  isFunction,
  isObjectOf({b: isString})
])"
`;

snapshot[`isIntersectionOf<T> > with symbol properties > returns properly named predicate function 1`] = `"isString"`;

snapshot[`isIntersectionOf<T> > with symbol properties > returns properly named predicate function 2`] = `
"isObjectOf({
  a: isNumber,
  Symbol(b): isString
})"
`;

snapshot[`isIntersectionOf<T> > with symbol properties > returns properly named predicate function 3`] = `
"isIntersectionOf([
  isFunction,
  isObjectOf({Symbol(b): isString})
])"
`;
